Initial projects charter (#54)
dotnet-foundation/projects: readme.md

# .NET Foundation Project Committee

## Background

The .NET Foundation, an independent 501(C)(6) corporation, is chartered to foster open development and collaboration around the growing collection of open source technologies for .NET. It serves as a forum for commercial and community developers alike to strengthen the future of the .NET ecosystem by promoting openness, community participation and rapid innovation.

> The mission of the .NET Foundation is to foster a community of .NET developers focused on collaborating on commercially friendly open source software projects that enhance the .NET platform and create value within the .NET ecosystem.

Reviewer: Should we add something about supporting smaller projects too so it doesn't sound like we just want only corporates to use the software. Eg commerical and hobbyist friendly.

Reviewer 2: The mission statement was copied from the original Advisory Council charter document.

Reviewer: Doesn't mean it can't be more inclusive. That's the point of pull requests.

Reviewer 2: Agree. I just wanted to clarify the origin of the mission statement included in this document.

Author: Resolving due to scope. The mission statement of the foundation can be addressed elsewhere.

## Rationale

An important aspect of the .NET Foundation's vision is to attract and foster vibrant communities of open source projects and contributors. However, in order to preserve the relevancy and credibility of the .NET Foundation's brand and reputation, it must ensure that all member projects fully embrace its mission and values and are willing to function as first class ambassadors and representatives of the .NET Foundation. It is the role of the Project Committee to assist the Board of Directors in ensuring this alignment.

## Responsibilities

The Project Committee manages the day to day requirements of Projects. This includes but is not limited to the following:

- Review and evaluate new Project applications
- Monitoring health of existing Projects
- Handling the retirement of Projects
- Providing support to Projects.

## Organizational Structure

> The Project Committee is comprised of volunteer .NET Foundation members and has a designated Chairperson appointed by the Executive Director. The Chairperson ensures that the Project Committee meets monthly to discuss and review items pertaining to Projects and provides information and recommendations to the Executive Director so that they can be presented to the Board of Directors for approval. All official communication from the .NET Foundation pertaining to Projects must be delivered by the Executive Director.

Reviewer 2: This was worded in an open manner because it may not always be possible that the Board has the capacity to lead all Foundation committees. .

> The Project Committee is comprised of volunteer .NET Foundation members and has a designated Chairperson appointed by the Executive Director. The Chairperson ensures that the Project Committee meets monthly to discuss and review items pertaining to Projects and provides information and recommendations to the Executive Director so that they can be presented to the Board of Directors for approval. All official communication from the .NET Foundation pertaining to Projects must be delivered by the Executive Director.

Author: I don't think the chairperson of each committee must be a board member, but I do think each committee should have at least one board member on it in an active capacity.

## Evaluation Criteria

> Transparency is a key open source characteristic, and as a result the criteria for evaluating .NET Foundation projects is designed to be as measurable and objective as possible. The criteria is not only important for evaluating new project applications but for measuring the ongoing health of existing projects. There are two types of criteria which have been defined:

Reviewer 3: _"The criteria **are** not only..."_?

**Eligibility** - this criterion is binary in nature and is used to determine if a project is fully aligned with the mission and values of the .NET Foundation.

> **Health** - this criterion is quantitative in nature and is based on publicly accessible metrics which can be used to evaluate the health and maturity of a project.

Reviewer 4: `Maturity` is still a little bit of a hang-up for me. It tends to be highly contextualized / debated, is easy to misconstrue, and almost always has some sort of organizational or communal politics attached to it.

As an alternative, I've been exploring the idea of confidence meters. And rather than specifying a requirement / target for such metrics, it's instead about transparency.

Confidence meters are transparent by nature. Confidence is from one group's perspective, but is not the law. Confidence is something that can be explained and others can choose to feel similarly based on the explanation.

Another key of a confidence meters: it doesn't make the Foundation responsible for getting the assessment correct, which also allows the foundation to be less prescriptive in its modeling.

So rather than health requirements, we could suggest that any .net member project will have a transparent confidence meters across consistent categories/areas.

Think about this manifesting itself on a scale of one to five, where five is very confident -- examples of rough draft items could include:

* The authors take appropriate security considerations and safeguard secrets and data as appropriate.
* Users will be aware of breaking changes in a standardized way.
* New changes to software will not endanger old functionality.
* There is a path for others to contribute to this project.
* The project is actively maintained

Etc.

Reviewer 2: Although the term "maturity" is used in a couple places in this document, the focus is on "health" - ie. there is Eligibility Criteria and Health Criteria.

## Project Levels

Projects are categorized according to the following levels: 

> **Applicant** - a project which has applied to the .NET Foundation to become a member. The Project Committee will review the application within one month of the submission to determine whether or not the project has satisfied the Eligibility and/or Health requirements. Any feedback to the applicant will be delivered by the Executive Director.

Author: The Project Committee should be entitled to comment on the application issue, ask questions, seek clarification, etc. Comm's don't need to go though the ED.

> **Seed** - a project which has been reviewed by the Project Committee and meets the basic Eligibility requirements but does not yet meet the Health requirements. Projects at this level are eligible for some .NET Foundation project services and will be re-evaluated automatically each quarter by the Project Committee to determine if they are ready to graduate to the Member level.

Reviewer 5: The monthly re-evaluation could become cumbersome if it's automatic.  I feel either the foundation should expend the effort to help move these Projects towards Member or cap the amount of re-evaluations.  Either is fine.  I just doubt we have the resources to re-evaluate projects that are clearly not moving towards Membership month after month indefinitely.

Reviewer 2: The criteria says that the re-evaluation will happen quarterly - not monthly - as monthly seemed too frequent.

Reviewer 2: @BillWagner this is a good suggestion to alleviate the burden on the Project Committee

**Member** - a project which satisfies the Eligibility and Health requirements and has been on-boarded as an official .NET Foundation project. Projects at this level can identify themselves publicly as Member Projects and are eligible to obtain access to all .NET Foundation project services. They will continue to be monitored on a monthly basis to ensure they meet the Eligibility and Health criteria.

> **Retired** - a project which was once a Member Project but no longer satisfies either the Eligibility or Health requirements, or has submitted a formal request to withdraw from the .NET Foundation.

Reviewer 6: This document may not be the place for it, but it could make sense for the Foundation to think through its processes and approach for when projects potentially no longer satisfy Foundation requirements. The Foundation would presumably want to communicate with projects about potential concerns before retiring a project.  Working in co-operation with projects to detail concerns where appropriate, and to support the project's continued health and membership, where possible, is presumably a sensible first goal before retiring a project.

## Eligibility Criteria

Broken into a variety of functional areas, the eligibility criteria is a checklist which defines if a project is fully aligned with the mission and values of the .NET Foundation.

**Suitability**

> - The project operates within moral and ethical standards of the .NET Foundation.

Reviewer 3: ```suggestion
- The project operates within the moral and ethical standards of the .NET Foundation.
```

- The project is aligned with the philosophy and guidelines for collaborative development.
- The project is built on the .NET platform and/or creates value within the .NET ecosystem.

> - The project solves a large enough problem and has a clear technology scope and focus which provides a reasonable likelihood of success.

Author: I think we can take this last point out and rely on the discretion of the committee and of the value statement in the previous bullet point.

**Code**

- The project produces software for distribution to the public at no charge.
- The project's code is easily discoverable and publicly accessible (preferably on GitHub). 
- The project contains a build script that can produce deployable artifacts that are identical to the official deployable artifacts, with the exception of code signing (Exception may be granted for strong name keys, though strongly encouraged to be committed. Exception relies on OSS signing being in the build script for public builds).
- When applicable, project must use reproducible build settings in its toolchain.

> - The project uses [Source Link](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink).

Reviewer 5: Does Source Link work for Xamarin platforms?  I doubt people are putting it in Xamarin projects if it doesn't function correctly on those target platforms.  Seems we should make sure it works across all .NET targets before we enforce it as a requirement?

Reviewer 7: Good question. I think we may want to make this a recommendation and not a requirement.

Author: AFAIK, Source Link works for all platforms. I strongly believe this should be a requirement as it establishes a link between the binary, the debugger and the source repo, which is critical towards supply chain security.

- The project uses either embedded PDBs or publish symbol packages to NuGet (if applicable).
- The project code signs their artifacts as appropriate.

**Licenses and Copyright**

- The code is offered under a standard, permissive open source license which has been approved by the .NET Foundation.
- Libraries that are mandatory dependencies of the project are also offered under a standard, permissive open source license which has been approved by the .NET Foundation (exceptions include a dependency on .NET Framework).

> - Committers are bound by a Contributor License Agreement.

Reviewer: May want to mention the license bots offered by the foundation here.

Author: Might be worth using an asterisk to denote requirements that can be met with Foundation services/support. Code Signing is another.

> - Committers are bound by a Contributor License Agreement.

Reviewer 3: Of what form does the CLA have to be?

Author: @martincostello not sure I understand the question? The foundation has a standard CLA we use for all projects.

Reviewer 3: Ah, ok. I took it to mean a project should have _a_ CLA.

Maybe include a link to a copy of the standard one used?

Reviewer 2: I think the challenge here is that a new project applying to be a member of the Foundation may have its own CLA - or maybe it has not used a CLA and does not have full provenance over its code.

Author: In the case where a CLA hasn't been used, which is the typical case, part of project onboarding, after approval/acceptance by the board, is for us to review the contributors who have made substantial contributions and include them on the IP transfer agreement.

We'd have to discuss the case of another CLA but we would have a strong preference to use ours to ensure that we can track things correctly.

- The copyright ownership of everything that the project produces is clearly defined and documented.

**Quality**

- The project is open and honest about the quality of its code.
- The project has a public issue tracker where the status of any defect can be easily obtained.
- The project puts a very high priority on producing secure software. 
- The project has a published Security Policy.

**Community**

> - The project has a home page which provides high level information about its status and purpose.

Reviewer 5: High level is subjective.  We should articulate minimum content for a website.

- The project welcomes contributions from anyone and embraces diversity in all forms.
- The project strives to be a meritocracy where more rights and responsibilities are granted to contributors who add value to the project.

> - The project has a public communication channel where community members can engage with maintainers.

Reviewer: I know some projects like prism have a patron based chat system rather than it being open to all. This could cause problems for some projects where the maintainers use a pay gate to avoid being overwhelmed.

- The project adheres to, and enforces, the .NET Foundation Code of Conduct.
- The project has a publicly available location where members can review documentation.

It is expected that a project satisfies the majority of the criteria outlined above; however, it is also possible for a project to obtain a waiver for specific criteria. It is the responsibility of the Project Committee to collate the project information, perform an evaluation, and make recommendations on a project's eligibility to the Board of Directors. The Board of Directors has the right to approve or deny a project's eligibility. Projects will also be re-evaluated periodically to ensure they continue to satisfy the eligibility requirements.

## Health Criteria

> Health criteria is designed to be measurable and is based on publicly accessible metrics which can be used to evaluate the health and maturity of a project on an ongoing basis. Metrics are based on values which can be readily obtained by anyone from GitHub and Nuget.

Reviewer 3: _"Health criteria are designed to be measurable and are based on..."_?

> Health criteria is designed to be measurable and is based on publicly accessible metrics which can be used to evaluate the health and maturity of a project on an ongoing basis. Metrics are based on values which can be readily obtained by anyone from GitHub and Nuget.

Reviewer 3: _"NuGet"_.

- Age (based on the date on the LICENSE file in the repository or a date provided by the maintainer) 
- Stars
- Watchers
- Forks

> - Used By

Reviewer 5: Used By seems to only be an internal metric to other projects on GitHub.  Is this a fair criteria if the code isn't sourced on GitHub?

Reviewer 8: To expand on that, are we restricting open-source projects to those hosted only on GitHub? Open source projects choosing to use an open source alternative (i.e. GitLab or similar) being locked out seems counterproductive.

Especially given that the Eligibility Criteria only *prefers* GitHub.

Reviewer 2: The goal of this proposal is for the Project Committee to be able to collect health metrics on a monthly basis for projects. Github makes these metrics publicly available, so there can be full transparency. That being said, if a project is not using Github there is the provision that the health criteria can be collected from other sources or waived altogether so that no projects are excluded based on where  they are hosted.

Reviewer 8: Will any of that information be documented? A GitLab-hosted project looking at this document would almost have to assume they'd be ineligible because they chose an open-source hosting platform, which is not a great impression.

Reviewer 2: The Eligibility criteria could be modified to say:

-The project's code is easily discoverable and publicly accessible

And the Health criteria could be modified to say:

For convenience reasons, standard metrics are based on values which can be readily obtained by anyone from GitHub and Nuget; however, metrics from other open source hosting platforms will also be considered.

- Commits
  - Issues Open
  - Issues Closed
  - Pull Requests Open

>   - Pull Requests Closed

Reviewer 9: On these four, I'd add a soft metric on aging issues and PRs. Issues, in particular, should be soft on resolution, but an active project should respond to suggestions with feedback in a reasonable time. PRs should get a review in a reasonable time.

- Branches
- Packages
- Releases
- Contributors

> - Nuget Package Downloads

Reviewer 3: ```suggestion
- NuGet Package Downloads
```

The Project Committee manages the day to day requirements of Projects. This includes but is not limited to the following:

- Review and evaluate new Project applications

> - Monitoring health of existing Projects

Reviewer 7: ```suggestion
- Monitor health of existing Projects
```

> - Handling the retirement of Projects

Reviewer 7: ```suggestion
- Handle the retirement of Projects
```

> - Providing support to Projects.

Reviewer 7: ```suggestion
- Provide support to Projects
```

Changed the language to be more active and consistent with the first entry.

**Suitability**

- The project operates within moral and ethical standards of the .NET Foundation.
- The project is aligned with the philosophy and guidelines for collaborative development.
- The project is built on the .NET platform and/or creates value within the .NET ecosystem.
- The project solves a large enough problem and has a clear technology scope and focus which provides a reasonable likelihood of success.

**Code**

> - The project produces software for distribution to the public at no charge.

Reviewer 7: Still have concerns about this. I believe the source code should be distributed at no charge per the[ Open Source Definition](https://opensource.org/osd-annotated), but we shouldn't require the software itself be free of charge. People should be able to charge for the end user software.

Reviewer 10: I also have concerns and believe the OSD isn't good enough. Thoughts:

- projects should be free to charge for distribution but must accept contributions from the community at the discretion of the project maintainers.

Ie "you are free to use the software, if the licensed allows (this against OSD) but ultimately we want to encourage you to engage with the project how it built in an open manner"

Ie. "Don't over index on the rights of the consumer"

Reviewer 7: > you are free to use the software, if the license allows (this against OSD)

I don't understand how this is against the OSD. Could you elaborate?

> Projects do not need to be use a OSD blessed license.

To be pedantic, the OSD is a definition that the open source community by and large accepts for a license to be considered open source. The OSI (Open Source Initiative) approves (and rejects) licenses.

So to restate your point, it seems you're suggesting that projects do not need to use a license that meets the Open Source Definition.

If that captures what you're suggesting, I would ask why should we allow projects to use a license that does not meet the OSD? I don't think the .NET Foundation wants to be in the business of defining its own open source definition. We don't have to accept _every_ license that meets the OSD. We could decide to only accept a subset. We could add extra requirements for .NET Foundation projects.

But I think the OSD is a good starting point unless there's some really compelling reason to diverge.

Reviewer 7: > projects should be free to charge for distribution but must accept contributions from the community at the discretion of the project maintainers who have a responsibility to act in the best interests of the project (which may just mean acting in the interests of the maintainers!)

This seems to contradict itself to be a non-statement. They __MUST__ accept contributions __AT THEIR DISCRETION__. Which is it? Must they, or do they have discretion?

In terms of evaluating a project, how do we evaluate this? Do we require that a project has accepted contributions from non-maintainers? What about projects that add all contributors to maintainers (some do this now).

Reviewer 10: @haacked I agree the OSD is a good starting point but don't agree that restricting scope of foundation to "opensource projects."

See @dependabot as a reference. It is NOT opensource as by the OSD. Yet it's openly developed. That's the situ I want to see more of.

What they are doing is totally valid and amazing but "is not opensource".

Look at all those PRs from the community fixing their own problems.

https://t.co/BiQQjFz8NQ

It hasn't been defined yet afaik . I'm calling it "openly developed software" - a wrapper around opensource that enables maintainers or projects to not be bound by the restrictions of OSI if they so choose. Key principles is openly developed.

That's what I'd like to see enabled and possible in the @dotnetfdn. It's all super new tho. Let's take this convo off GitHub as I don't want to distract folks. See twitter DM.

Reviewer 11: Even though we both have very similar _thoughts_ on open software, I'm going to deviate from @ghuntley a bit here. I'm a big fan of supporting and normalizing "openly developed" software as an alternative to "open source" (conceding "open source" == OSI-approved license). There are lots of good reasons why a maintainer might choose to go that route and it should be their choice to make.

That said, I actually think it's okay for the Foundation to take the position that "openly developed software is cool and all, but that's just not our wheelhouse for projects inside the .NET Foundation." It _doesn't_ mean that the Foundation can't support, promote, or even use (I.e. Dependabot) such projects in other ways and I would hope that it does, just that those projects probably aren't suitable to be _member projects_.

...which opens up a bigger question for me about "what really is a member project?" - I'll open a separate comment for that.

- The project's code is easily discoverable and publicly accessible (preferably on GitHub). 
- The project contains a build script that can produce deployable artifacts that are identical to the official deployable artifacts, with the exception of code signing (Exception may be granted for strong name keys, though strongly encouraged to be committed. Exception relies on OSS signing being in the build script for public builds).
- When applicable, project must use reproducible build settings in its toolchain.
- The project uses [Source Link](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink).
- The project uses either embedded PDBs or publish symbol packages to NuGet (if applicable).
- The project code signs their artifacts as appropriate.

**Licenses and Copyright**

> - The code is offered under a standard, permissive open source license which has been approved by the .NET Foundation.

Reviewer 7: Where can people find a list of licenses approved by the .NET Foundation? We should have a link to the exact list if we plan to curate allowed permissive licenses.

Reviewer 10: I agree with list of licensed but disagree with granting this much power to the OSI.

See https://github.com/dotnet-foundation/projects/pull/54#discussion_r382249773 for discussion.

> - Libraries that are mandatory dependencies of the project are also offered under a standard, permissive open source license which has been approved by the .NET Foundation (exceptions include a dependency on .NET Framework).

Reviewer 7: What about the situation where a project targets proprietary hardware and such hardware requires a dependency on a proprietary license?

Reviewer 10: See also https://github.com/dotnet-foundation/projects/pull/54#discussion_r382249773

Reviewer 7: Maybe I'm being dense, but I didn't see how https://github.com/dotnet-foundation/projects/pull/54#discussion_r382249773 addressed my concern here.

Imagine this scenario: You have a .NET Lib that makes it easy to add hand gestures for the Oculus Rift. But it turns out you have to take a mandatory dependency on _oculus-rift.dll_ which is proprietary. The way this is worded right now, we would not allow such a project.

I think we should. I think we can make this a bit more generic.

```suggestion
- Libraries that are mandatory dependencies of the project are also offered under a standard, permissive open source license which has been approved by the .NET Foundation (exceptions include a dependency that is required by the target platform where no alternative open source dependency is available such as the .NET Framework or a hardware specific library).
```

- Committers are bound by a Contributor License Agreement.
- The copyright ownership of everything that the project produces is clearly defined and documented.

**Quality**

- The project is open and honest about the quality of its code.
- The project has a public issue tracker where the status of any defect can be easily obtained.
- The project puts a very high priority on producing secure software. 

> - The project has a published Security Policy.

Reviewer 7: We should have a link to a Security Policy template and some examples.

**Community**

- The project has a home page which provides high level information about its status and purpose.
- The project welcomes contributions from anyone and embraces diversity in all forms.

> - The project strives to be a meritocracy where more rights and responsibilities are granted to contributors who add value to the project.

Reviewer 7: Strongly feel we should not embrace the word "meritocracy" here. It was [coined as satire](https://kottke.org/17/03/the-satirical-origins-of-the-meritocracy) and does not mean what we intend it to mean here to many people. In fact, it can be a divisive word: https://readwrite.com/2014/01/24/github-meritocracy-rug/

Reviewer 2: I was not aware of the origin of the word and I have never seen it used in a divisive manner. In fact, I see it regularly used in open source communities and it seems that the majority of people are familiar with the more traditional definition of the word - ie. https://www.merriam-webster.com/dictionary/meritocracy or https://en.wikipedia.org/wiki/Meritocracy

Reviewer 7: @sbwalker the wikipedia article you referenced has a whole section of Criticism on the term meritocracy which begins with...

> The term "meritocracy" was originally intended as a negative concept.

And continues to highlight some of the issues with the term. The term meritocracy is often used to push out members of marginalized groups. There's [a discussion in the Geek Feminism Wiki ](https://geekfeminism.wikia.org/wiki/Meritocracy) that cites several criticisms of the concept.
